File: src/services/messengers/wa.js
"use strict"
const downloader = require("../../utils/download-services");
const mediaConverter = require("../../utils/media-converters");
const fileServices = require("../../utils/file-services");
const { insertMessage } = require("../messages/messages-service");
const axios = require("axios");

class EventKindE {
  static STATUS_UPDATE = 'status_update';
  static MESSAGE = 'message';
}

class MessageKindE {
  static TEXT = 'text';
  static VOICE = 'voice'; // a voice recorded by the app
  static AUDIO = 'audio'; // a file attachement of type audio
  // there are many more kinds
}

function getEventKind(value) {
  if (value.hasOwnProperty('statuses')) return EventKindE.STATUS_UPDATE;
  if (value.hasOwnProperty('messages')) return EventKindE.MESSAGE;
  // shouldn't happen
  return undefined;
}

function getMessageKind(value) {
  switch(value.type) {
    //case 'audio': return (value.audio.voice) ? MessageKindE.VOICE : MessageKindE.AUDIO;
    case 'audio': return MessageKindE.VOICE;
    default: return value.type; // just lazy, as all other texts match.
  }
}

function parseMessage(message) {
  const source = "wa";

  const eventKind = getEventKind(message.entry[0].changes[0].value);
  if (eventKind != EventKindE.MESSAGE) {
    return undefined;
  }

  const kind = getMessageKind(message.entry[0].changes[0].value.messages[0]);

  // User messages that come from WA have int timestamp in [sec]
  // WA doesn't generate an event for our reply messages, so need to generate manually.
  // For manually generated timestamps we use date.now() which responds with [ms] as int.
  // To keep the accuracy, and get the same [sec] units we do date.now()/1000, and pass it as float.
  // So here, parseFloat works both for WA int timestamps and our float timestamps.
  const messageTimestamp = parseFloat(message.entry[0].changes[0].value.messages[0].timestamp)* 1e3;
  const senderId = message.entry[0].changes[0].value.messages[0].from;

  // TODO igors - the current code assumes responses to chatID (the TG approach)
  // in WA the response is to a phone number.
  // WA has a chatId at    message.entry[0].id  but it seems to be meaningless.
  // so for now, have chatId carry the person to respond to
  const chatId = senderId;

  // WhatsApp Business API does not allow for group chats.
  const chatType = "private";
  const isSentByMe = senderId == process.env.WHATSAPP_PHONE_NUMBER;
  const messageId = message.entry[0].changes[0].value.messages[0].id;
  const replyToMessageId = message.entry[0].changes[0].value.messages[0].hasOwnProperty('context') ?
    message.entry[0].changes[0].value.messages[0].context.id:
    undefined;
  
  const body = (kind == MessageKindE.TEXT) ? message.entry[0].changes[0].value.messages[0].text.body : undefined;
  const fileId = (kind == MessageKindE.VOICE) ? message.entry[0].changes[0].value.messages[0].audio.id : undefined;
  const fileUniqueId = undefined;

  return [{
    source,
    messageTimestamp,
    chatType,
    chatId,
    senderId,
    isSentByMe,
    messageId,
    replyToMessageId,
    kind,
    body,
    rawSource: message
  }, {
    fileId,
    fileUniqueId
  }];
}

function getBotGeneratedMessage(ctx, sendMessageResponse, attributes) {
  // This represents a skeleton of the message that would be received by a webhook on the client side
  // if the cline would be running a webhook.
  const { chatId, quoteId, kind, body } = attributes;
  const message = {
    "entry": [
      {
        "changes": [
          {
            "value": {
              "messages": [
                {
                  "timestamp": (Date.now()/1e3).toString(), // WA sends time in sec, so normalize to that
                  "from": process.env.WHATSAPP_PHONE_NUMBER.toString(),
                  "id": sendMessageResponse.data.messages[0].id,
                  "type": kind,
                  "text": {
                    "body": body
                  }
                }
              ]
            }
          }
        ]
      }
    ]
  };

  return message;
}

async function sendMessage(ctx, attributes) {
  const { chatId, quoteId, kind, body } = attributes;
  const response = await sendMessageRaw(ctx, attributes);

  if (response.hasOwnProperty('data')) {
    // for now (text messages) the existance of data should be enough to indicate success.
    // WA doesn't return a message struct that is received by the client, so no info to use to call parse message.
    // for now, just build a fake message
    const message = getBotGeneratedMessage(ctx, response, attributes);

    // TODO ishumsky - fileInfo is outside until added to the DB.
    const [parsedMessage, fileInfo] = parseMessage(message);

    // TODO igors - because of the abuse of chatId, after parsing it'll have the value of our bot,
    // so DB lookups will not pick responses.
    // change back to the user phone number to correlate.
    parsedMessage.chatId = chatId;
    ctx.log({ parsedMessage });

    await insertMessage(ctx, parsedMessage);
    ctx.log(`Message inserted successfully: `, parsedMessage);

  }
}

async function sendMessageRaw(ctx, attributes) {
  const { chatId, quoteId, kind, body } = attributes;

  if (kind != "text") {
    return;
  }

  const headers = {
    Authorization: `Bearer ${process.env.WHATSAPP_BOT_TOKEN}`,
    'Content-Type': 'application/json'
  };

  if (body.length > 4000) { body = body.substring(0, 4000); }

  var args = {
    messaging_product: "whatsapp",    
    recipient_type: "individual",
    to: chatId,
    type: "text",
    text: {
      preview_url: false,
      body: body
    }
  };

  if (quoteId) {
    args.context = {message_id: quoteId};
  }

  let response;
  try {
     response = await axios.post(
      `https://graph.facebook.com/${process.env.FACEBOOK_GRAPH_VERSION}/${process.env.WHATSAPP_PHONE_NUMBER_ID}/messages`,
      args,
      {headers}
    );
  } catch (error) {
    ctx.log(`sendMessageRaw: exception. error.response.data=${error?.response?.data}`);
    if (error?.response?.data) {
      ctx.log(JSON.stringify(error?.response?.data));
    }
    throw(error);
  }
  //ctx.log(response);
  
  return response;
}

function isMessageForMe(msg) {
  if (msg.chatType == "private") {
    return true;
  }

  return false;
}

async function getVoiceMp3File(ctx, parsedMessage, fileInfo) {
  ctx.log(`getVoiceMp3File: ${parsedMessage}, ${fileInfo}`);
  const url = await getDownloadUrl(ctx, fileInfo.fileId);
  const [oggFilePath, mp3FilePath] = getAudioFilePaths(ctx, parsedMessage.chatId, fileInfo);
  let isDownloadSuccessful = false;
  try {
    const headers = {
      Authorization: `Bearer ${process.env.WHATSAPP_BOT_TOKEN}`,
    };

    isDownloadSuccessful = await downloader.downloadStreamFile(ctx, url, oggFilePath, headers);
    await mediaConverter.convertOggToMp3(ctx, oggFilePath, mp3FilePath);

    return mp3FilePath;
    
  } finally {
    // we should delete the Ogg file no matter what happened, as long as it exists.
    const deleteOggFile = isDownloadSuccessful || fileServices.fileExists(oggFilePath);
    if (deleteOggFile) {
      fileServices.deleteFile(ctx, oggFilePath);
    }
  }
}

async function getDownloadUrl(ctx, fileId) {
  ctx.log(`getDownloadUrl: ${fileId}`);
  const headers = {
    Authorization: `Bearer ${process.env.WHATSAPP_BOT_TOKEN}`,
  };

  let response;
  try {
    response = await axios.get(
      `https://graph.facebook.com/${process.env.FACEBOOK_GRAPH_VERSION}/${fileId}?phone_number_id=${process.env.WHATSAPP_PHONE_NUMBER_ID}`,
      {headers}
    );
  } catch (error) {
    ctx.log(`getDownloadUrl: exception. error.response.data=${error?.response?.data}`);
    throw(error);
  }

  if (response.hasOwnProperty("error")) {
    ctx.log('getDownloadUrl failed. response=', response);
  }

  ctx.log(`getDownloadUrl: response=${response}`);
  // from now on assume it has succeeeded.

  const downloadUrl = response.data.url;

  ctx.log(`getDownloadUrl: downloadUrl=${downloadUrl}`);
  return downloadUrl;
}

function getAudioFilePaths(ctx, chatId, fileInfo) {
  const tempDirPath = fileServices.makeTempDirName(`r1x/wa/${chatId}_`);
  const filePathName = tempDirPath + '/audio';
  const oggFilePath = filePathName + '.ogg';
  const mp3FilePath = filePathName + '.mp3';

  ctx.log(`getAudioFilePaths: oggFilePath=${oggFilePath}, mp3FilePath=${mp3FilePath}`);
  return [oggFilePath, mp3FilePath];
}


function setTyping(chatId, inFlight) {
  // TODO igors - can't find WA API for typing indication.
  return;
}

async function setStatusRead(ctx, messageId) {
  ctx.log(`setStatusRead`);
  const headers = {
    Authorization: `Bearer ${process.env.WHATSAPP_BOT_TOKEN}`,
    'Content-Type': 'application/json',
  };

  var args = {
    messaging_product: "whatsapp",
    status: "read",
    message_id: messageId,
  };

  let response;
  try {
    response = await axios.post(
      `https://graph.facebook.com/${process.env.FACEBOOK_GRAPH_VERSION}/${process.env.WHATSAPP_PHONE_NUMBER_ID}/messages`,
      args,
      {headers}
    );
  } catch (error) {
    ctx.log(`setStatusRead: exception. error.response.data=${error?.response?.data}`);
    return ;
  }

  if (response?.data?.success != true) {
    ctx.log(`setStatusRead: response is wrong. Compared field ${response?.data?.success}. Full response ${response}`);
  }
}

module.exports = {
  parseMessage,
  sendMessage,
  sendMessageRaw,
  isMessageForMe,
  setTyping,
  getVoiceMp3File,
  setStatusRead
};

File: src/services/messengers/wa.py
import os
from typing import Dict
import requests
from services.messengers.messenger import MessageKindE, MessagingService
from infra import utils
from box import Box
import time

from infra.context import Context

class EventKindE:
    STATUS_UPDATE = 'status_update'
    MESSAGE = 'message'
    

class WhatsappMessenger(MessagingService):
    def _get_event_kind(self, value):
        if 'statuses' in value:
            return EventKindE.STATUS_UPDATE
        if 'messages' in value:
            return EventKindE.MESSAGE
        return None

    def _get_message_kind(self, value) -> str:
        if value['type'] == 'audio':
            return MessageKindE.VOICE
        return value['type']

    def parse_message(self, message):
        source = "wa"
        event_kind = self._get_event_kind(message['entry'][0]['changes'][0]['value'])
        if event_kind != EventKindE.MESSAGE:
            return None

        message0 = message['entry'][0]['changes'][0]['value']['messages'][0]

        kind = self._get_message_kind(message0)
        message_timestamp = float(message0['timestamp'])
        sender_id = self.chat_id
        chat_type = "private"
        is_sent_by_me = sender_id == os.environ['WHATSAPP_PHONE_NUMBER']
        is_forwarded = (message0.get('context', {}).get('forwarded', None) != None)
        message_id = message0['id']
        reply_to_message_id = message0.get('context', {}).get('id')

        if kind == MessageKindE.TEXT:
            body = message0['text']['body']
        else:
            body = None

        if kind == MessageKindE.VOICE:
            file_id = message0['audio']['id']
        else:
            file_id = None

        file_unique_id = None

        return [Box({
            "source": source,
            "messageTimestamp": message_timestamp,
            "chatType": chat_type,
            "chatId": self.chat_id,
            "senderId": sender_id,
            "isSentByMe": is_sent_by_me,
            "isForwarded" : is_forwarded,
            "messageId": message_id,
            "replyToMessageId": reply_to_message_id,
            "kind": kind,
            "body": body,
            "rawSource": message
        }), Box({
            "fileId": file_id,
            "fileUniqueId": file_unique_id
        })]

    def _get_bot_generated_message(self, ctx:Context, send_message_response, attributes):
        quote_id = attributes.get('quote_id')
        kind = attributes.get('kind')
        body = attributes.get('body')

        message = {
            "entry": [
                {
                    "changes": [
                        {
                            "value": {
                                "messages": [
                                    {
                                        "timestamp": (int(time.time() * 1000) / 1e3),
                                        "from": os.environ['WHATSAPP_PHONE_NUMBER'],
                                        "id": send_message_response['messages'][0]['id'],
                                        "type": kind,
                                        "text": {
                                            "body": body
                                        }
                                    }
                                ]
                            }
                        }
                    ]
                }
            ]
        }

        return message

    def send_message(self, ctx:Context, attributes):
        quote_id = attributes.get('quote_id')
        kind = attributes.get('kind')
        body = attributes.get('body')

        if kind != "text":
            return

        if len(body) > 4000:
            ctx.log('send_message: message body too long, %d > 4000' % len(body))
            body = body[0:3999]

        headers = {
            "Authorization": f"Bearer {os.environ['WHATSAPP_BOT_TOKEN']}",
            "Content-Type": "application/json"
        }

        args = {
            "messaging_product": "whatsapp",
            "recipient_type": "individual",
            "to": self.chat_id,
            "type": "text",
            "text": {
                "preview_url": False,
                "body": body
            }
        }

        if quote_id:
            args["context"] = {"message_id": quote_id}

        response = self._post_message_request(ctx, headers, args)

        if response == None:
            return None

        message = self._get_bot_generated_message(ctx, response.json(), attributes)
        parsed_message, _ = self.parse_message(message)
        parsed_message.chatId = self.chat_id

        return parsed_message

    def _post_message_request(self, ctx:Context, headers:Dict[str,str], args):
        try:
            response = requests.post(
                f"https://graph.facebook.com/{os.environ['FACEBOOK_GRAPH_VERSION']}/{os.environ['WHATSAPP_PHONE_NUMBER_ID']}/messages",
                json=args,
                headers=headers
            )
            response.raise_for_status()
        except requests.exceptions.RequestException as error:
            ctx.log(f"post_message_request: exception. error={error}")
            raise error
        return response
    
    def send_contact(self, ctx: Context, name:str, handle:str):
        headers = {
            "Authorization": f"Bearer {os.environ['WHATSAPP_BOT_TOKEN']}",
            "Content-Type": "application/json"
        }
        contact_args = {
            "messaging_product": "whatsapp",
            "recipient_type": "individual",
            "to": self.chat_id,
            "type": "contacts",
            "contacts": [
                {
                    "addresses": [],
                    "emails": [],
                    "name": {
                        "first_name": name,
                        "formatted_name": name,
                        "last_name": ""
                    },
                    "org": {},
                    "phones": [
                        {
                            "phone": f"+{handle}",
                            "type": "HOME",
                            "wa_id": handle
                        }
                    ],
                    "urls": []
                }
            ]
        }
        response = self._post_message_request(ctx,headers,contact_args)
        return response.json()     

    def is_message_for_me(self, msg) -> bool:
        if msg.chatType == "private":
            return True

        return False

    def get_voice_mp3_file(self, ctx:Context, parsed_message, file_info, work_dir) -> str:
        ctx.log(f"getVoiceMp3File: {parsed_message}, {file_info}, {work_dir}")
        url = self._get_download_url(ctx, file_info.fileId)
        orig_file_path, mp3_file_path = self._get_audio_file_paths(ctx, parsed_message.chatId, file_info, work_dir)

        headers = {
            "Authorization": f"Bearer {os.environ['WHATSAPP_BOT_TOKEN']}",
        }

        utils.download_stream_file(ctx, url, orig_file_path, headers)
        utils.convert_audio_to_mp3(ctx, orig_file_path, mp3_file_path)

        return mp3_file_path

    def _get_download_url(self, ctx:Context, file_id):
        ctx.log(f"getDownloadUrl: {file_id}")
        headers = {
            "Authorization": f"Bearer {os.environ['WHATSAPP_BOT_TOKEN']}",
        }

        try:
            response = requests.get(
                f"https://graph.facebook.com/{os.environ['FACEBOOK_GRAPH_VERSION']}/{file_id}?phone_number_id={os.environ['WHATSAPP_PHONE_NUMBER_ID']}",
                headers=headers
            )
            response.raise_for_status()
        except requests.exceptions.RequestException as error:
            ctx.log(f"getDownloadUrl: exception. error={error}")
            raise error

        download_url = response.json()['url']

        ctx.log(f"getDownloadUrl: downloadUrl={download_url}")
        return download_url

    def _get_audio_file_paths(self, ctx:Context, chat_id, file_info, work_dir):
        orig_file_path = work_dir / 'audio.orig'
        mp3_file_path = work_dir / 'audio.mp3'

        ctx.log(f"getAudioFilePaths: orgFilePath={orig_file_path}, mp3FilePath={mp3_file_path}")

        return orig_file_path, mp3_file_path


    def set_typing(self, in_flight):
        # TODO: igors - can't find WA API for typing indication.
        pass


    def set_status_read(self, ctx:Context, message_id):
        ctx.log("setStatusRead")
        headers = {
            "Authorization": f"Bearer {os.environ['WHATSAPP_BOT_TOKEN']}",
            "Content-Type": "application/json",
        }

        args = {
            "messaging_product": "whatsapp",
            "status": "read",
            "message_id": message_id,
        }

        try:
            response = requests.post(
                f"https://graph.facebook.com/{os.environ['FACEBOOK_GRAPH_VERSION']}/{os.environ['WHATSAPP_PHONE_NUMBER_ID']}/messages",
                json=args,
                headers=headers
            )
            response.raise_for_status()
        except requests.exceptions.RequestException as error:
            ctx.log(f"setStatusRead: exception. error={error}")
            return

        if response.json().get('success') != True:
            ctx.log(f"setStatusRead: response is wrong. Compared field {response.json().get('success')}. Full response {response}")
